File: tests/unit/StrategyYVault/test_set_admin.py
import brownie

ZERO_ADDRESS = "0x0000000000000000000000000000000000000000"


def test_set_admin(accounts, strategyYVault):
    tx = strategyYVault.setAdmin(accounts[1], { 'from': accounts[0] })

    assert strategyYVault.admin() == accounts[1]
    assert len(tx.events) == 1
    assert tx.events["SetAdmin"].values() == [accounts[1]]


def test_set_admin_not_admin(accounts, strategyYVault):
    with brownie.reverts("!admin"):
        strategyYVault.setAdmin(accounts[1], {'from': accounts[1]})


def test_set_admin_to_zero_address(accounts, strategyYVault):
    with brownie.reverts("admin == zero address"):
        strategyYVault.setAdmin(ZERO_ADDRESS, {'from': accounts[0]})

File: test/unit/ControllerV2/test-set-admin.ts
import chai from "chai"
import { ControllerV2Instance } from "../../../types"
import { ZERO_ADDRESS } from "../../util"
import _setup from "./setup"

contract("Controller", (accounts) => {
  const refs = _setup(accounts)
  const { admin } = refs

  let controller: ControllerV2Instance
  beforeEach(() => {
    controller = refs.controller
  })

  describe("setAdmin", () => {
    it("should set admin", async () => {
      await controller.setAdmin(accounts[1], { from: admin })

      assert.equal(await controller.admin(), accounts[1])

      const ADMIN_ROLE = await controller.ADMIN_ROLE()
      const HARVESTER_ROLE = await controller.HARVESTER_ROLE()

      assert.equal(
        await controller.hasRole(ADMIN_ROLE, admin),
        false,
        "revoke admin role"
      )
      assert.equal(
        await controller.hasRole(HARVESTER_ROLE, admin),
        false,
        "revoke harvester role"
      )
      assert.equal(
        await controller.hasRole(ADMIN_ROLE, accounts[1]),
        true,
        "grant admin role"
      )
      assert.equal(
        await controller.hasRole(HARVESTER_ROLE, accounts[1]),
        true,
        "grant harvester role"
      )
    })

    it("should reject if caller not admin", async () => {
      await chai
        .expect(controller.setAdmin(accounts[1], { from: accounts[1] }))
        .to.be.rejectedWith("!admin")
    })

    it("should reject zero address", async () => {
      await chai
        .expect(controller.setAdmin(ZERO_ADDRESS, { from: admin }))
        .to.be.rejectedWith("admin = zero address")
    })
  })
})
